Mathlib/LinearAlgebra/Matrix/Irreducible/Defs.lean

```diff
@@ -0,0 +1,308 @@
+`[DecidableEq n]` to reason about finite sums.
+
+## References
+
+*   [E. Seneta, *Non-negative Matrices and Markov Chains*][seneta2006]
+
+## Tags
+
+matrix, nonnegative, positive, power, quiver, graph, irreducible, primitive, perron-frobenius
+-/
+
+namespace Matrix
+
+open Finset Quiver Quiver.Path
+
+variable {n R : Type*} [Fintype n] [Ring R] [LinearOrder R]
+
+@[simp]
+lemma pow_nonneg [IsOrderedRing R] [DecidableEq n] {A : Matrix n n R}
+    (hA : ∀ i j, 0 ≤ A i j) (k : ℕ) : ∀ i j, 0 ≤ (A ^ k) i j := by
+  induction k with
+  | zero => intro i j; simp [one_apply]; by_cases h : i = j <;> simp [h]
+  | succ m ih =>
+    intro i j; rw [pow_succ, mul_apply]
+    exact Finset.sum_nonneg fun l _ => mul_nonneg (ih i l) (hA l j)
+
+/-- The directed graph (quiver) associated with a matrix `A`,
+with an edge `i ⟶ j` iff `0 < A i j`. -/
+def toQuiver (A : Matrix n n R) : Quiver n :=
+  ⟨fun i j => 0 < A i j⟩
+
+/-- A matrix `A` is irreducible if it is entrywise nonnegative and
+its quiver of positive entries (`toQuiver A`) is strongly connected. -/
+def Irreducible (A : Matrix n n R) : Prop :=
+  (∀ i j, 0 ≤ A i j) ∧ @IsSStronglyConnected n (toQuiver A)
+
+/-- A matrix `A` is primitive if it is entrywise nonnegative
+and some positive power has all entries strictly positive. -/
+def IsPrimitive [DecidableEq n] (A : Matrix n n R) : Prop :=
+  (∀ i j, 0 ≤ A i j) ∧ ∃ k, 0 < k ∧ ∀ i j, 0 < (A ^ k) i j
+
+variable {A : Matrix n n R}
+
+/-- If `A` is irreducible and `n>1` then every row has a positive entry. -/
+lemma Irreducible.no_zero_row
+    (h_irr : Irreducible A) (h_dim : 1 < Fintype.card n) (i : n) :
+    ∃ j, 0 < A i j := by
+  letI : Quiver n := toQuiver A
+  by_contra h_row
+  have no_out : ∀ j : n, IsEmpty (i ⟶ j) :=
+    fun j => ⟨fun e => h_row ⟨j, e⟩⟩
+  obtain ⟨j, hij⟩ := Fintype.exists_ne_of_one_lt_card h_dim i
+  obtain ⟨p, hp_pos⟩ := h_irr.2 i j
+  have h_le : 1 ≤ p.length := Nat.succ_le_of_lt hp_pos
+  have ⟨v, p₁, p₂, _hp_eq, hp₁_len⟩ := p.exists_eq_comp_of_le_length (n := 1) h_le
+  have hlen_ne : p₁.length ≠ 0 := by simp [hp₁_len]
+  obtain ⟨c, p', e, rfl⟩ :=
+    (Quiver.Path.length_ne_zero_iff_eq_cons (p := p₁)).1 hlen_ne
+  have hp'_succ : p'.length + 1 = 1 := by
+    simpa [Quiver.Path.length_cons] using hp₁_len
+  have hp'0 : p'.length = 0 := by
+    exact Nat.succ.inj hp'_succ
+  have hi_c : i = c := Quiver.Path.eq_of_length_zero p' hp'0
+  subst hi_c
+  exact (no_out _).false e
+
+lemma sum_pos_of_mem [IsOrderedRing R] {α : Type*} {s : Finset α} {f : α → R}
+    (h_nonneg : ∀ a ∈ s, 0 ≤ f a) (a : α) (ha_mem : a ∈ s) (ha_pos : 0 < f a) :
+    0 < ∑ x ∈ s, f x := by
+  classical
+  rw [Eq.symm (add_sum_erase s f ha_mem),  ]
+  have h_erase_nonneg : 0 ≤ ∑ x ∈ s.erase a, f x :=
+    Finset.sum_nonneg (fun x hx => h_nonneg x (Finset.mem_of_mem_erase hx))
+  rw [add_sum_erase s f ha_mem]; rw [propext (sum_pos_iff_of_nonneg h_nonneg)];
+  exact Filter.frequently_principal.mp fun a_1 ↦ a_1 ha_mem ha_pos
```

Comment:
This is a good lemma but it already exists (with slightly weaker assumptions) as [Finset.sum_pos'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Order/BigOperators/Group/Finset.html#Finset.sum_pos')
```suggestion
```